clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java

```diff
@@ -25,75 +29,173 @@
-            return MemoryRecords.withLogEntries(compressionType, converted);
+            // maintain the batch along with the decompressed records to avoid the need to decompress again
+            List<RecordBatchAndRecords> recordBatchAndRecordsList = new ArrayList<>(batches.size());
+            int totalSizeEstimate = 0;
+
+            for (RecordBatch batch : batches) {
+                if (batch.magic() <= toMagic) {
+                    totalSizeEstimate += batch.sizeInBytes();
+                    recordBatchAndRecordsList.add(new RecordBatchAndRecords(batch, null, null));
+                } else {
+                    List<Record> records = Utils.toList(batch.iterator());
+                    final long baseOffset;
+                    if (batch.magic() >= RecordBatch.MAGIC_VALUE_V2)
+                        baseOffset = batch.baseOffset();
+                    else
+                        baseOffset = records.get(0).offset();
+                    totalSizeEstimate += estimateSizeInBytes(toMagic, baseOffset, batch.compressionType(), records);
+                    recordBatchAndRecordsList.add(new RecordBatchAndRecords(batch, records, baseOffset));
+                }
+            }
+
+            ByteBuffer buffer = ByteBuffer.allocate(totalSizeEstimate);
+            for (RecordBatchAndRecords recordBatchAndRecords : recordBatchAndRecordsList) {
+                if (recordBatchAndRecords.batch.magic() <= toMagic)
+                    recordBatchAndRecords.batch.writeTo(buffer);
+                else
+                    buffer = convertLogEntry(toMagic, buffer, recordBatchAndRecords);
+            }
+
+            buffer.flip();
+            return MemoryRecords.readableRecords(buffer);
         }
     }
 
-    public static int estimatedSize(CompressionType compressionType, Iterable<LogEntry> entries) {
-        int size = 0;
-        for (LogEntry entry : entries)
-            size += entry.sizeInBytes();
-        // NOTE: 1024 is the minimum block size for snappy encoding
-        return compressionType == CompressionType.NONE ? size : Math.min(Math.max(size / 2, 1024), 1 << 16);
+    private ByteBuffer convertLogEntry(byte magic, ByteBuffer buffer, RecordBatchAndRecords recordBatchAndRecords) {
+        RecordBatch batch = recordBatchAndRecords.batch;
+        final TimestampType timestampType = batch.timestampType();
+        long logAppendTime = timestampType == TimestampType.LOG_APPEND_TIME ? batch.maxTimestamp() : RecordBatch.NO_TIMESTAMP;
+
+        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, magic, batch.compressionType(),
+                timestampType, recordBatchAndRecords.baseOffset, logAppendTime);
+        for (Record record : recordBatchAndRecords.records) {
+            // control messages are only supported in v2 and above, so skip when down-converting
+            if (magic < RecordBatch.MAGIC_VALUE_V2 && record.isControlRecord())
+                continue;
```

Comment:
With this, it's possible to have no records added to the builder. Does the builder support that at close() time?